near-sdk/src/collections/tree_map.rs

```diff
@@ -0,0 +1,1772 @@
+        test_env::setup();
+        let map: TreeMap<u32, u32> = TreeMap::default();
+        let _ = map.range((Bound::Excluded(2), Bound::Included(1)));
+    }
+
+    #[test]
+    #[should_panic(expected = "Invalid range.")]
+    fn test_range_panics_non_overlap_incl_incl() {
+        test_env::setup();
+        let map: TreeMap<u32, u32> = TreeMap::default();
+        let _ = map.range((Bound::Included(2), Bound::Included(1)));
+    }
+
+    #[test]
+    fn test_iter_rev_from_empty() {
+        test_env::setup();
+        let map: TreeMap<u32, u32> = TreeMap::default();
+        assert!(map.iter_rev_from(42).collect::<Vec<(u32, u32)>>().is_empty());
+    }
+
+    //
+    // Property-based tests of AVL-based TreeMap against std::collections::BTreeMap
+    //
+
+    fn avl<K, V>(insert: &[(K, V)], remove: &[K]) -> TreeMap<K, V>
+        where
+            K: Ord + Copy + BorshSerialize + BorshDeserialize,
+            V: Default + Copy + BorshSerialize + BorshDeserialize,
+    {
+        test_env::setup_free();
+        let mut map: TreeMap<K, V> = TreeMap::default();
+        for k in remove {
+            map.insert(k, &Default::default());
+        }
+        let n = insert.len().max(remove.len());
+        for i in 0..n {
+            if i < remove.len() {
+                map.remove(&remove[i]);
+            }
+            if i < insert.len() {
+                let (k, v) = insert[i];
+                map.insert(&k, &v);
+            }
+        }
+        map
+    }
+
+    fn rb<K, V>(insert: &[(K, V)], remove: &[K]) -> BTreeMap<K, V>
+        where
+            K: Ord + Copy + BorshSerialize + BorshDeserialize,
+            V: Default + Copy + BorshSerialize + BorshDeserialize,
+    {
+        let mut map: BTreeMap<K, V> = BTreeMap::default();
+        for k in remove {
+            map.insert(*k, Default::default());
+        }
+        let n = insert.len().max(remove.len());
+        for i in 0..n {
+            if i < remove.len() {
+                map.remove(&remove[i]);
+            }
+            if i < insert.len() {
+                let (k, v) = insert[i];
+                map.insert(k, v);
+            }
+        }
+        map
+    }
+
+    #[test]
+    fn prop_avl_vs_rb() {
+        fn prop(insert: Vec<(u32, u32)>, remove: Vec<u32>) -> bool {
+            let a = avl(&insert, &remove);
+            let b = rb(&insert, &remove);
+            let v1: Vec<(u32, u32)> = a.iter().collect();
+            let v2: Vec<(u32, u32)> = b.into_iter().collect();
+            v1 == v2
+        }
+
+        QuickCheck::new()
+            .tests(300)
+            .quickcheck(prop as fn(std::vec::Vec<(u32, u32)>, std::vec::Vec<u32>) -> bool);
+    }
+
+    #[test]
+    fn prop_avl_height() {
+        test_env::setup_free();
+
+        fn prop(insert: Vec<(u32, u32)>, remove: Vec<u32>) -> bool {
+            let map = avl(&insert, &remove);
+            map.height() <= max_tree_height(map.len())
+        }
```

Comment:
I add this check to this test:
```    
    fn check_all_balances<K, V>(map: &TreeMap<K, V>, root: u64)
    where
        K: Ord + Copy + BorshSerialize + BorshDeserialize,
        V: Copy + BorshSerialize + BorshDeserialize,
    {
        if let Some(node) = map.node(root) {
            let balance = map.get_balance(&node);
            assert!(-1 <= balance && balance <= 1, "BAD BALANCE: balance {}, node id={}, ht={}, l={:?}, r={:?}", balance, node.id, node.ht, node.lft, node.rgt);
            if let Some(id) = node.lft {
                map.check_all_balances(id);
            }
            if let Some(id) = node.rgt {
                map.check_all_balances(id);
            }
        }
    }

    #[test]
    fn prop_avl_height() {
        test_env::setup_free();

        fn prop(insert: Vec<(u32, u32)>, remove: Vec<u32>) -> bool {
            let map = avl(&insert, &remove);
            check_all_balances(&map, map.root);
            map.height() <= max_tree_height(map.len())
        }

        QuickCheck::new()
            .tests(300)
            .quickcheck(prop as fn(std::vec::Vec<(u32, u32)>, std::vec::Vec<u32>) -> bool);
    }
```
and it fails, is my assert wrong?

Reply:
looking into it

Reply:
Assertion `-1 <= balance <= 1` describes perfectly balanced tree (at each level), height of such tree would always be `log(N+1)`. So assertion is not wrong, just too strict for AVL tree.

AVL gives no guarantees of being perfectly balanced at each level after each insert/remove, thus `log(N+1)` is a lower bound of tree height, and upper bound is: `C * log2(n + D) + B` according to [wiki](https://en.wikipedia.org/wiki/AVL_tree).

Reply:
Quote from wiki:
```
of its two child sub-trees. A binary tree is defined to be an AVL tree if the invariant

    BalanceFactor ( n o d e ) ∈ { − 1 , 0 , 1 } {\displaystyle {\text{BalanceFactor}}(node)\in {\{-1,0,1\}}} {\displaystyle {\text{BalanceFactor}}(node)\in {\{-1,0,1\}}}[7]

holds for every n o d e {\displaystyle node} {\displaystyle node} in the tree. 
```

Reply:
still looking into it then

Reply:
`remove` indeed does not restore AVL-tree invariants, investigating further

Reply:
Thank you, @mikhailOK ! I was wrong relying only on tree height to validate balance, as well as about definition of a perfectly balanced tree (it is only perfectly balanced when balance == 0 in all nodes), and also about AVL-tree fundamental property - balance among all subtrees.

Fixed now, `prop_avl_balance` passes.